### PR TITLE
Refactor nested loop in getTPMapMatchingSpreadConstraints

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
In getTPMapMatchingSpreadConstraints, the loop over constraints appears three times in processNode func.

This PR changes the order of nested loop (loop over constraints versus loop over pods) and saves traversal over constraints.

Motivation for this PR is that, it seems constraints play an important role for this func. The refactored code is more understandable and achieves efficiency at the same time.

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 1]
```diff
@@ -230,22 +230,20 @@ func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*sche
...
+		for _, constraint := range constraints {
+			// Ensure current node's labels contains all topologyKeys in 'constraints'.
+			if _, ok := node.Labels[constraint.TopologyKey]; !ok {
+				return
 			}
-			podLabelSet := labels.Set(existingPod.Labels)
-			for _, constraint := range constraints {
-				ok, err := podMatchesSpreadConstraint(podLabelSet, constraint)
+			pairAdded := false
+			for _, existingPod := range nodeInfo.Pods() {
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/metadata.go]: If current `node.Labels` match on the 1st constraint's `topologyKey` **only**, we're wasting a large number of cycles on this loop, correct?

Comment (tedyu) [reply pkg/scheduler/algorithm/predicates/metadata.go]: When you say 'current node.Labels', did you mean existingPod.Labels ?

How would existing code be able to short circuit and avoid the waste ?

Thanks

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: No, I mean node.Labels.

Suppose you have 2 constraints, the 2nd constraint's topology isn't present in the node's labels.
- Without this PR, it directly returns without processing any pod.
- With this PR, it has to process nodeInfo.Pods() and only return when it hits the 2nd constraint.

Comment (tedyu) [reply pkg/scheduler/algorithm/predicates/metadata.go]: I restored the first loop over constraints.

**Comment by tedyu** [comment]
@Huang-Wei 
Mind taking another look ?

thanks

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 2]
```diff
@@ -230,22 +230,22 @@ func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*sche
...
+
 		// Ensure current node's labels contains all topologyKeys in 'constraints'.
 		for _, constraint := range constraints {
 			if _, ok := node.Labels[constraint.TopologyKey]; !ok {
 				return
 			}
 		}
-
 		nodeTopologyMaps := newTopologyPairsMaps()
 		// nodeInfo.Pods() can be empty; or all pods don't fit
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/metadata.go]: Please move the line above `for _, existingPod :=`

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 3]
```diff
@@ -254,21 +254,20 @@ func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*sche
...
-			// constraint.TopologyKey is already guaranteed to be present
-			pair := topologyPair{
-				key:   constraint.TopologyKey,
-				value: node.Labels[constraint.TopologyKey],
+			if !pairAdded {
+				pair := topologyPair{
+					key:   constraint.TopologyKey,
+					value: node.Labels[constraint.TopologyKey],
+				}
+				// If needed, append topology pair without entry of pods.
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/metadata.go]: Please move L265~L268 above L260 `if !pairAdde`.

**Review by Huang-Wei — COMMENTED** [review]
Just some nits, otherwise LGTM.

**Comment by tedyu** [comment]
@Huang-Wei 
Your comment has been addressed.

**Comment by Huang-Wei** [comment]
/lgtm
/approve

Thanks, @tedyu!

**Pull request merged**